src/crypto/cose.cpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+
+#include "ccf/crypto/cose.h"
+
+#include <optional>
+#include <qcbor/qcbor_decode.h>
+#include <qcbor/qcbor_encode.h>
+#include <qcbor/qcbor_spiffy_decode.h>
+#include <stdexcept>
+
+namespace ccf::cose::edit
+{
+  std::vector<uint8_t> set_unprotected_header(
+    const std::span<const uint8_t>& cose_input,
+    int64_t key,
+    pos::Type pos,
+    const std::vector<uint8_t> value)
+  {
+    UsefulBufC buf{cose_input.data(), cose_input.size()};
+
+    QCBORError err;
+    QCBORDecodeContext ctx;
+    QCBORDecode_Init(&ctx, buf, QCBOR_DECODE_MODE_NORMAL);
+
+    size_t pos_start = 0;
+    size_t pos_end = 0;
+
+    QCBORDecode_EnterArray(&ctx, nullptr);
+    err = QCBORDecode_GetError(&ctx);
+    if (err != QCBOR_SUCCESS)
+    {
+      throw std::logic_error("Failed to parse COSE_Sign1 outer array");
+    }
+
+    auto tag = QCBORDecode_GetNthTagOfLast(&ctx, 0);
+    if (tag != CBOR_TAG_COSE_SIGN1)
+    {
+      throw std::logic_error("Failed to parse COSE_Sign1 tag");
+    }
+
+    QCBORItem item;
+    err = QCBORDecode_GetNext(&ctx, &item);
+    if (err != QCBOR_SUCCESS || item.uDataType != QCBOR_TYPE_BYTE_STRING)
+    {
+      throw std::logic_error(
+        "Failed to parse COSE_Sign1 protected header as bstr");
+    }
+    UsefulBufC phdr = {item.val.string.ptr, item.val.string.len};
+
+    // Skip unprotected header
+    QCBORDecode_VGetNextConsume(&ctx, &item);
+
+    err = QCBORDecode_PartialFinish(&ctx, &pos_start);
+    if (err != QCBOR_ERR_ARRAY_OR_MAP_UNCONSUMED)
+    {
+      throw std::logic_error("Failed to find start of payload");
+    }
+    QCBORDecode_VGetNextConsume(&ctx, &item);
+    err = QCBORDecode_PartialFinish(&ctx, &pos_end);
+    if (err != QCBOR_ERR_ARRAY_OR_MAP_UNCONSUMED)
+    {
+      throw std::logic_error("Failed to find end of payload");
+    }
+    UsefulBufC payload = {cose_input.data() + pos_start, pos_end - pos_start};
```

Comment:
I wonder, does it work properly for detached payload or shall we detect that separately and set `QCBOREncode_AddNULL`?

Reply:
I also wonder, RFC says

```payload:  This field contains the serialized content to be signed.
      If the payload is not present in the message, the application is
      required to supply the payload separately.  The payload is wrapped
      in a bstr to ensure that it is transported without changes.  If
      the payload is transported separately ("detached content"), then a
      nil CBOR object is placed in this location, and it is the
      responsibility of the application to ensure that it will be
      transported without changes.
```

So can we just expect `bstr` type only instead of playing around with `...PartialDecode`?

Reply:
It does, see tests. The payload may or may not be a bstr.